### PR TITLE
Fix quirks mode regression

### PR DESCRIPTION
**Pull request by zm-cttae**
- We have a significant regression so we need to ~~revert~~ handle #100.
  The only other method of fixing this one is `srcdoc` and that's included in the Trusted Types policy.
- The raster image methods were self-cleaning since #71.
  Per #70 comments, let's run it in `toSvg` to make sure sandbox is killed in all cases.

### CODE REVIEW

File: src/dom-to-image-more.js [hunk 1]
```diff
@@ -1199,13 +1199,30 @@
...
+                    isTrustedTypesRequired = csp.indexOf('require-trusted-types-for') !== -1;
+                } catch(_) {
+                    isTrustedTypesRequired = true;
+                }
+                if (!isTrustedTypesRequired) {
+                    const sandboxDocument = document.implementation.createHTMLDocument(sandbox.id);
+                    const sandboxDoctype = document.doctype ? '<!DOCTYPE html>' : '';
+                    const sandboxHTML = `${sandboxDoctype}${sandboxDocument.documentElement.outerHTML}`;
+                    sandbox.setAttribute('srcdoc', sandboxHTML);
+                }
```
Comment (meche-gh) [inline src/dom-to-image-more.js:1224]: Thoughts @joswhite?

File: src/dom-to-image-more.js [hunk 2]
```diff
@@ -89,6 +89,7 @@
 
         function clearCache(result) {
             domtoimage.impl.urlCache = [];
+            removeSandbox();
```
Comment (IDisposable) [inline src/dom-to-image-more.js:92]: I'm going to do this change in a separate PR just for neatness.

File: src/dom-to-image-more.js (hunk 1 above)
Comment (meche-gh) [reply src/dom-to-image-more.js:1224]: Forgot to restore the charset in L1220-1223.. will fix.

Comment (meche-gh) [reply src/dom-to-image-more.js]: Fixed in 1ac5c78

File: src/dom-to-image-more.js (hunk 2 above)
Comment (IDisposable) [reply src/dom-to-image-more.js:92]: ... and released that change in both v2.14.1 and v3.0.2-rc.0

Comment (IDisposable) [reply src/dom-to-image-more.js:92]: ... and WOO! as of this weekend, we're getting picked up by CDNJS :) https://cdnjs.com/libraries/dom-to-image-more

File: src/dom-to-image-more.js [hunk 3]
```diff
@@ -1199,13 +1199,31 @@
...
+                // Ensure the iframe's character rendering matches the document (with UTF-8 fallback).
                 const charset = document.createElement('meta');
                 charset.setAttribute('charset', document.characterSet || 'UTF-8');
                 sandbox.contentDocument.head.appendChild(charset);
-                sandbox.contentDocument.title = 'sandbox';
+                // Ensure the document is rendered in standard mode by coercing doctype.
+                // If the parent document has a Trusted Types policy, fallback to quirks mode.
+                let isTrustedTypesRequired;
+                try {
+                    const xhr = new XMLHttpRequest();
```
Comment (IDisposable) [inline src/dom-to-image-more.js:1211]: Is there no better way to determine this _for the currently loaded page_ without reloading it this way? Seems like this will be very slow generating an entire round-trip + server response time...

Comment (IDisposable) [reply src/dom-to-image-more.js:1211]: We may have to let people opt-in to this...

Comment (meche-gh) [reply src/dom-to-image-more.js:1211]: I think it wouldn't be as bad if it was fetching the `HEAD` only instead of `GET`.  
Even with the artificial 2000ms delay that "Slow 3G" throttling adds, its a consistent 100ms.

File: src/dom-to-image-more.js [hunk 4]
```diff
@@ -1199,13 +1199,31 @@
...
+                } catch(_) {
+                    isTrustedTypesRequired = true;
+                }
+                if (!isTrustedTypesRequired) {
+                    const sandboxDocument = document.implementation.createHTMLDocument(sandbox.id);
+                    sandboxDocument.head.appendChild(charset);
+                    const sandboxDoctype = document.doctype ? '<!DOCTYPE html>' : '';
+                    const sandboxHTML = `${sandboxDoctype}${sandboxDocument.documentElement.outerHTML}`;
+                    sandbox.setAttribute('srcdoc', sandboxHTML);
+                }
```
Comment (IDisposable) [inline src/dom-to-image-more.js:1225]: So, the question I have is what does this actually do versus the original code? Seems like the path where we don't need trusted types is handled now, but what happens if we _do need them_?

**Review by IDisposable — COMMENTED** [review]
Is it possible to create test cases for each possibility?

File: src/dom-to-image-more.js (hunk 4 above)
Comment (meche-gh) [reply src/dom-to-image-more.js:1225]: If the page would have thrown an error because of Trusted Types, it'll have the `iframe` with the #100 regression present. If it doesn't have a Trusted Types policy - most pages thankfully don't, we do go ahead and coerce in the doctype.

File: src/dom-to-image-more.js (hunk 3 above)
Comment (meche-gh) [reply src/dom-to-image-more.js]: Resolved in 6b609d3

File: src/dom-to-image-more.js [hunk 5]
```diff
@@ -1199,13 +1199,31 @@
                 sandbox.style.visibility = 'hidden';
                 sandbox.style.position = 'fixed';
                 document.body.appendChild(sandbox);
-                // Ensure that the iframe is rendered in standard mode
+                sandbox.id = sandbox.contentDocument.title = 'domtoimage-sandbox';
```
Comment (IDisposable) [inline src/dom-to-image-more.js:1202]: Worried that we should be using our `uid()` method to synthesize a unique ID (to be used inside here only, obviously) instead of "hoping" it won't clash. This would follow the pattern for the class names we create.

File: src/dom-to-image-more.js (hunk 4 above)
Comment (IDisposable) [reply src/dom-to-image-more.js:1225]: Seems icky to be doing the DOCTYPE handling only in one case, but I guess because we're creating the IFRAME from the main document, it gets the right one inside?

File: src/dom-to-image-more.js (hunk 5 above)
Comment (meche-gh) [reply src/dom-to-image-more.js]: Resolved in fd264d9

File: src/dom-to-image-more.js (hunk 1 above)
Comment (meche-gh) [reply src/dom-to-image-more.js]: We should be shipshape after Marc's review work :) resolving

File: src/dom-to-image-more.js (hunk 4 above)
Comment (meche-gh) [reply src/dom-to-image-more.js:1225]: Yep. HTML is the correct doctype for our default style query seeing as we embed inside a HTML `foreignObject`

Comment (IDisposable) [reply src/dom-to-image-more.js:1225]: My thought (so poorly expressed, sorry), is that the net effect of this code when the `isTrustedTypesRequired === true` is that we're not actually setting the DOCTYPE of the new `IFRAME` at all. I am wondering if we can/should... seems that the DOM `.doctype` is read-only, so I doubt we can change anything since we're already synthesized the` iframe` up at line 1198. I guess this PR (and the one that led to it) has gone through enough churn that I am a bit gunshy about this whole code segment. Sorry.

I might be a bit less worried if we were doing _anything_ in the (non-existent) `else` clause for this test... in that my mind could settle on what's supposed to be happening in either case, but I'm too close.

File: src/dom-to-image-more.js [hunk 6]
```diff
@@ -1220,7 +1220,7 @@
                     const sandboxDocument = document.implementation.createHTMLDocument(sandbox.id);
                     sandboxDocument.head.appendChild(charset);
                     const sandboxDoctype = document.doctype ? '<!DOCTYPE html>' : '';
-                    const sandboxHTML = `${sandboxDoctype}${sandboxDocument.documentElement.outerHTML}`;
+                    const sandboxHTML = sandboxDoctype + sandboxDocument.documentElement.outerHTML;
```
Comment (IDisposable) [inline src/dom-to-image-more.js:1223]: Much better... :)

File: src/dom-to-image-more.js [hunk 7]
```diff
@@ -1199,7 +1199,7 @@
                 sandbox.style.visibility = 'hidden';
                 sandbox.style.position = 'fixed';
                 document.body.appendChild(sandbox);
-                sandbox.id = sandbox.contentDocument.title = 'domtoimage-sandbox';
+                sandbox.id = sandbox.contentDocument.title = 'domtoimage-sandbox-' + util.uid();
```
Comment (IDisposable) [inline src/dom-to-image-more.js:1202]: Good :)

File: src/dom-to-image-more.js [hunk 8]
```diff
@@ -1199,13 +1199,31 @@
...
-                sandbox.contentDocument.title = 'sandbox';
+                // Ensure the document is rendered in standard mode by coercing doctype.
+                // If the parent document has a Trusted Types policy, fallback to quirks mode.
+                let isTrustedTypesRequired;
+                try {
+                    const request = new XMLHttpRequest();
+                    request.open('HEAD', document.location.href, false);
+                    request.send();
+                    const csp = request.getResponseHeader('content-security-policy');
+                    isTrustedTypesRequired = csp.indexOf('require-trusted-types-for') !== -1;
```
Comment (joswhite) [inline src/dom-to-image-more.js:1215]: To check for Trusted Type API error, instead of creating a network request, what if we try the original `sandbox.contentWindow.document.write` in the `try` block? Then in the `catch` block, we execute the fallback code.

Another idea - would `documeny.write` succeed if we used `DOMPurity.sanitize` first? See an example [here](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/require-trusted-types-for#examples)

Comment (joswhite) [reply src/dom-to-image-more.js:1215]: One more thing - using XMLHttpRequest is deprecated. I get the following warning in my console - "Synchronous XMLHttpRequest on the main thread is deprecated because of its detrimental effects to the end user’s experience. For more help http://xhr.spec.whatwg.org/"

Comment (IDisposable) [reply src/dom-to-image-more.js:1215]: Deep into this rabbit-hole... stumbled upon this comment https://github.com/w3c/trusted-types/blob/main/src/trustedtypes.js#L337 which worried me a bit, but I think that's only Edge 12-18 (e.g. the not-Chromium-based version) https://caniuse.com/iframe-srcdoc 

Noting this here so I don't head down the same rabbit hole.

Comment (IDisposable) [reply src/dom-to-image-more.js:1215]: Looks like we could just look to see if there's a CSP on the currently executing script like https://github.com/w3c/trusted-types/blob/main/src/polyfill/full.js#L27

Comment (M00TSDNS) [reply src/dom-to-image-more.js:1215]: 😄 seeing as we are in the rabbit hole ...

Try-finally doesn't work for some strange reason. However we can actually use the Trusted Types API to check if a policy is needed here - [spec](https://w3c.github.io/trusted-types/dist/spec/#dom-trustedtypepolicyfactory-getpropertytype)

Comment (meche-gh) [reply src/dom-to-image-more.js:1215]: 😄 seeing as we are in the rabbit hole ...

try-finally doesn't work for some strange reason. However we can actually use the Trusted Types API to check if a policy is needed here - [spec](https://w3c.github.io/trusted-types/dist/spec/#dom-trustedtypepolicyfactory-getpropertytype)

Comment (IDisposable) [reply src/dom-to-image-more.js:1215]: Trying to get somewhere , how does this look?
```javascript
        function ensureSandboxWindow() {
            if (!sandbox) {
                // figure out how this document is defined (doctype and charset)
                const charsetToUse = document.characterSet || 'UTF-8';               
                const docType = document.doctype;
                const docTypeDeclaration = docType
                    ? `<!DOCTYPE ${docType.name} ${docType.publicId} ${docType.systemId}`.trim() + '>'
                    : '';

                // Create a hidden sandbox <iframe> element within we can create default HTML elements and query their
                // computed styles. Elements must be rendered in order to query their computed styles. The <iframe> won't
                // render at all with `display: none`, so we have to use `visibility: hidden` with `position: fixed`.
                sandbox = document.createElement('iframe');
                sandbox.id = 'domtoimage-sandbox-' + util.uid();
                sandbox.style.visibility = 'hidden';
                sandbox.style.position = 'fixed';
                document.body.appendChild(sandbox);

                tryTechniques(sandbox, docTypeDeclaration, charsetToUse, 'domtoimage-sandbox');
            }

            return sandbox.contentWindow;
        }

        function tryTechniques(sandbox, doctype, charset, title) {
            // try the good old-fashioned document write with all the correct attributes set
            try {
                sandbox.contentWindow.document.write(`${doctype}<html><head><meta charset='${charset}'><title>${title}</title></head><body></body></html>`);
                return sandbox.contentWindow;
            }
            catch (_) {
                // swallow exception and fall through to next technique
            }

            const metaCharset = document.createElement('meta');
            metaCharset.setAttribute('charset', charset);

            // let's attempt it using srcdoc, so we can still set the doctype and charset
            try {
                const sandboxDocument = document.implementation.createHTMLDocument(title);
                sandboxDocument.head.appendChild(metaCharset);
                const sandboxHTML = doctype + sandboxDocument.documentElement.outerHTML;
                sandbox.setAttribute('srcdoc', sandboxHTML);
                return sandbox.contentWindow;
            } catch (_) {
                // swallow exception and fall through to the simplest path
            }

            // let's attempt it using srcdoc... here we're not able to set the doctype
            sandbox.contentDocument.head.appendChild(metaCharset);
            sandbox.contentDocument.title = title;
            return sandbox.contentWindow;
        }

```

Comment (IDisposable) [reply src/dom-to-image-more.js:1215]: I've pushed that as `quirks-regression` branch https://github.com/1904labs/dom-to-image-more/tree/quirks-regression

Can you validate @joswhite as I don't have a fiddle or unit test that shows this.

Comment (meche-gh) [reply src/dom-to-image-more.js:1215]:
```diff
 - let's attempt it using srcdoc
 + let's attempt it using contentDocument
```

Comment (IDisposable) [reply src/dom-to-image-more.js:1215]: Bummer...

image

doesn't look much like

image


```console
await domtoimage.toSvg(document.body)
VM183:1221 
        
       This document requires 'TrustedHTML' assignment.
tryTechniques @ VM183:1221
ensureSandboxWindow @ VM183:1212
getDefaultStyle @ VM183:1151
copyUserComputedStyleFast @ VM183:1071
copyStyle @ VM183:367
cloneStyle @ VM183:342
Promise.then (async)
processClone @ VM183:333
(anonymous) @ VM183:266
Promise.then (async)
cloneNode @ VM183:265
(anonymous) @ VM183:294
Promise.then (async)
(anonymous) @ VM183:293
cloneChildren @ VM183:292
(anonymous) @ VM183:263
Promise.then (async)
cloneNode @ VM183:262
(anonymous) @ VM183:294
Promise.then (async)
(anonymous) @ VM183:293
cloneChildren @ VM183:292
(anonymous) @ VM183:263
Promise.then (async)
cloneNode @ VM183:262
(anonymous) @ VM183:294
Promise.then (async)
(anonymous) @ VM183:293
cloneChildren @ VM183:292
(anonymous) @ VM183:263
Promise.then (async)
cloneNode @ VM183:262
(anonymous) @ VM183:76
Promise.then (async)
toSvg @ VM183:75
(anonymous) @ VM188:1
VM183:1236 
        
       This document requires 'TrustedHTML' assignment.
tryTechniques @ VM183:1236
ensureSandboxWindow @ VM183:1212
getDefaultStyle @ VM183:1151
copyUserComputedStyleFast @ VM183:1071
copyStyle @ VM183:367
cloneStyle @ VM183:342
Promise.then (async)
processClone @ VM183:333
(anonymous) @ VM183:266
Promise.then (async)
cloneNode @ VM183:265
(anonymous) @ VM183:294
Promise.then (async)
(anonymous) @ VM183:293
cloneChildren @ VM183:292
(anonymous) @ VM183:263
Promise.then (async)
cloneNode @ VM183:262
(anonymous) @ VM183:294
Promise.then (async)
(anonymous) @ VM183:293
cloneChildren @ VM183:292
(anonymous) @ VM183:263
Promise.then (async)
cloneNode @ VM183:262
(anonymous) @ VM183:294
Promise.then (async)
(anonymous) @ VM183:293
cloneChildren @ VM183:292
(anonymous) @ VM183:263
Promise.then (async)
cloneNode @ VM183:262
(anonymous) @ VM183:76
Promise.then (async)
toSvg @ VM183:75
(anonymous) @ VM188:1
callout:1 
        
       Access to XMLHttpRequest at 'https://www.google.com/images/hpp/Chrome_Owned_96x96.png' from origin 'https://ogs.google.com' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.
VM183:764 
        
       cannot fetch resource: https://www.google.com/images/hpp/Chrome_Owned_96x96.png, status: 0
fail @ VM183:764
done @ VM183:738
XMLHttpRequest.send (async)
(anonymous) @ VM183:719
getAndEncode @ VM183:707
Promise.then (async)
inline @ VM183:997
(anonymous) @ VM183:1016
Promise.then (async)
inlineAll @ VM183:1014
(anonymous) @ VM183:1020
(anonymous) @ VM183:1019
Promise.then (async)
inlineAll @ VM183:1014
(anonymous) @ VM183:1020
(anonymous) @ VM183:1019
Promise.then (async)
inlineAll @ VM183:1014
(anonymous) @ VM183:1020
(anonymous) @ VM183:1019
Promise.then (async)
inlineAll @ VM183:1014
(anonymous) @ VM183:1020
(anonymous) @ VM183:1019
Promise.then (async)
inlineAll @ VM183:1014
(anonymous) @ VM183:1020
(anonymous) @ VM183:1019
Promise.then (async)
inlineAll @ VM183:1014
(anonymous) @ VM183:1020
(anonymous) @ VM183:1019
Promise.then (async)
inlineAll @ VM183:1014
(anonymous) @ VM183:1020
(anonymous) @ VM183:1019
Promise.then (async)
inlineAll @ VM183:1014
(anonymous) @ VM183:1020
(anonymous) @ VM183:1019
Promise.then (async)
inlineAll @ VM183:1014
(anonymous) @ VM183:1020
(anonymous) @ VM183:1019
Promise.then (async)
inlineAll @ VM183:1014
(anonymous) @ VM183:1020
(anonymous) @ VM183:1019
Promise.then (async)
inlineAll @ VM183:1014
inlineImages @ VM183:483
Promise.then (async)
toSvg @ VM183:79
(anonymous) @ VM188:1
VM183:719 
        
      
        
        
      
        
      
       
        
       GET https://www.google.com/images/hpp/Chrome_Owned_96x96.png net::ERR_FAILED 200
```

Comment (meche-gh) [reply src/dom-to-image-more.js:1215]: The Trusted Types API fn only tells us if we _could_ be hitting the problem, it isn't a requirement check.  
Looks like async XHR with HEAD method is the only sane check we could use here.

Comment (meche-gh) [reply src/dom-to-image-more.js:1215]: The API `detectPolicy` snippet doesn't handle trusted type policy that is set as response header on server-side. :(  
Async XHR with HEAD method it is..

Comment (IDisposable) [reply src/dom-to-image-more.js:1215]: Sorry, this has been so painful. Thanks for the patience.

Comment (joswhite) [reply src/dom-to-image-more.js:1215]: > I've pushed that as `quirks-regression` branch https://github.com/1904labs/dom-to-image-more/tree/quirks-regression and it runs all the unit tests correctly
> 
> Can you validate @joswhite as I don't have a fiddle or unit test that shows this.

I haven't been able to prove that the old code actually activated Quirks mode. For other `<iframe>` instances without DOCTYPE, I get this. Not sure why I couldn't get the old code to show me the same warning. It might depend on the environment or browser as well.

image

Anyways, I didn't see any obvious issues. Thanks so much for your work on this!
